go/proxyd/cache.go

```diff
@@ -0,0 +1,275 @@
+	if val, ok := c.lru.Get(key); ok {
+		return val.(string), nil
+	}
+	return "", nil
+}
+
+func (c *cache) Put(ctx context.Context, key string, value string) error {
+	c.lru.Add(key, value)
+	return nil
+}
+
+type redisCache struct {
+	rdb *redis.Client
+}
+
+func newRedisCache(url string) (*redisCache, error) {
+	opts, err := redis.ParseURL(url)
+	if err != nil {
+		return nil, err
+	}
+	rdb := redis.NewClient(opts)
+	if err := rdb.Ping(context.Background()).Err(); err != nil {
+		return nil, wrapErr(err, "error connecting to redis")
+	}
+	return &redisCache{rdb}, nil
+}
+
+func (c *redisCache) Get(ctx context.Context, key string) (string, error) {
+	val, err := c.rdb.Get(ctx, key).Result()
+	if err == redis.Nil {
+		return "", nil
+	} else if err != nil {
+		return "", err
+	}
+	return val, nil
+}
+
+func (c *redisCache) Put(ctx context.Context, key string, value string) error {
+	err := c.rdb.Set(ctx, key, value, 0).Err()
+	return err
+}
+
+type GetLatestBlockNumFn func(ctx context.Context) (uint64, error)
+
+type RPCCache struct {
+	cache               Cache
+	getLatestBlockNumFn GetLatestBlockNumFn
+}
+
+func newRPCCache(cache Cache, getLatestBlockNumFn GetLatestBlockNumFn) *RPCCache {
+	return &RPCCache{cache: cache, getLatestBlockNumFn: getLatestBlockNumFn}
+}
+
+func (c *RPCCache) GetRPC(ctx context.Context, req *RPCReq) (*RPCRes, error) {
+	if !c.isCacheable(req) {
+		return nil, nil
+	}
+
+	key := mustMarshalJSON(req)
+	encodedVal, err := c.cache.Get(ctx, string(key))
+	if err != nil {
+		return nil, err
+	}
+	if encodedVal == "" {
+		return nil, nil
+	}
+
+	val, err := snappy.Decode(nil, []byte(encodedVal))
+	if err != nil {
+		return nil, err
+	}
+
+	res := new(RPCRes)
+	err = json.Unmarshal(val, res)
+	if err != nil {
+		panic(err)
+	}
+
+	return res, nil
+}
+
+func (c *RPCCache) PutRPC(ctx context.Context, req *RPCReq, res *RPCRes) error {
+	if !c.isCacheable(req) {
+		return nil
+	}
+	if supportedBlockRPCMethods[req.Method] {
+		if ok, err := c.isConfirmed(ctx, req); err != nil {
+			return err
+		} else if !ok {
+			return nil
+		}
+	}
+
+	key := mustMarshalJSON(req)
+	val := mustMarshalJSON(res)
+	encodedVal := snappy.Encode(nil, val)
+	return c.cache.Put(ctx, string(key), string(encodedVal))
+}
+
+func (c *RPCCache) isConfirmed(ctx context.Context, req *RPCReq) (bool, error) {
+	if c.getLatestBlockNumFn == nil {
+		return false, errUnavailableBlockNumSyncer
+	}
+	curBlock, err := c.getLatestBlockNumFn(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	switch req.Method {
+	case "eth_getBlockByNumber":
+		blockInput, _, err := decodeGetBlockByNumberParams(req.Params)
+		if err != nil {
+			return false, err
+		}
+		if isBlockDependentParam(blockInput) {
+			return false, nil
+		}
+		if blockInput == "earliest" {
+			return true, nil
+		}
+		blockNum, err := decodeBlockInput(blockInput)
+		if err != nil {
+			return false, err
+		}
+		return blockNum+numBlockConfirmations <= curBlock, nil
+
+	case "eth_getBlockRange":
```

Comment:
Just a note: this is a custom RPC method that isn't in upstream geth and will be removed when we move to the next version of the system